controller/orphan_controller.go

```diff
@@ -269,48 +303,187 @@ func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if closeErr := imc.Close(); closeErr != nil {
+			oc.logger.WithError(closeErr).Error("failed to close instance manager client")
+		}
+	}()
+
+	runtimeInstanceExist := types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeRuntimeInstanceState).Status == longhorn.ConditionStatusTrue
+	if runtimeInstanceExist {
+		if err = oc.deleteRuntimeInstance(imc, instance, imType, deType); err != nil {
+			return false, err
+		}
+	}
+	return oc.confirmOrphanRuntimeInstanceCleanup(imc, instance, imType, deType)
+}
+
+func (oc *OrphanController) extractOrphanedRuntimeInfo(orphan *longhorn.Orphan) (name string, dataEngineType longhorn.DataEngineType, err error) {
+	name, ok := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get instance name for runtime instance orphan %v", orphan.Name)
+	}
+
+	dataEngineTypeStr, ok := orphan.Spec.Parameters[longhorn.OrphanDataEngineType]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get data engine type for runtime instance orphan %v", orphan.Name)
+	}
+	switch dataEngineType = longhorn.DataEngineType(dataEngineTypeStr); dataEngineType {
+	case longhorn.DataEngineTypeV1, longhorn.DataEngineTypeV2:
+		// supported data engine type
 	default:
-		err = fmt.Errorf("unknown orphan type %v", orphan.Spec.Type)
+		return "", "", fmt.Errorf("unknown data engine type %v for runtime instance orphan %v", dataEngineTypeStr, orphan.Name)
 	}
 
-	if err == nil || (err != nil && apierrors.IsNotFound(err)) {
-		return nil
+	return name, dataEngineType, nil
+}
+
+func (oc *OrphanController) deleteRuntimeInstance(imc *engineapi.InstanceManagerClient, instanceName string, instanceKind longhorn.InstanceManagerType, engineType longhorn.DataEngineType) (err error) {
+	// There is a delay between deletion initiation and state/InstanceManager update,
+	// this function may be called multiple times before given instance exits.
+
+	oc.logger.Infof("Deleting runtime instance %v", instanceName)
+
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot delete runtime instance %v", instanceName)
+		}
+	}()
+
+	err = imc.InstanceDelete(engineType, instanceName, string(instanceKind), "", false)
+	if err != nil && !types.ErrorIsNotFound(err) {
+		return err
 	}
 
-	return err
+	return nil
+}
+
+func (oc *OrphanController) confirmOrphanRuntimeInstanceCleanup(imc *engineapi.InstanceManagerClient, instanceName string, imType longhorn.InstanceManagerType, engineType longhorn.DataEngineType) (isCleanupComplete bool, err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot confirm cleanup result of runtime instance %v", instanceName)
+		}
+	}()
+
+	_, err = imc.InstanceGet(engineType, instanceName, string(imType))
+	isCleanupComplete = false
+	if err != nil && types.ErrorIsNotFound(err) {
+		isCleanupComplete = true
+		err = nil
+	}
+	return isCleanupComplete, err
```

Comment:
NIT: consider tweaking this for better readabiliy and maintainablity.

```suggestion
	_, err = imc.InstanceGet(engineType, instanceName, string(imType))
	switch {
	case err == nil:
		return false, nil // instance still exists - cleanup not complete.
	case types.ErrorIsNotFound(err):
		return true, nil // instance not found - cleanup completed.
	default:
		return false, err // Unexpected error - cleanup status unknown.
	}
```